File: operating system I/dining_philosophers_problem.py
#!/usr/bin/env python3

from __future__ import print_function
from threading import Thread, Lock
from time import sleep
import random

__description__ = "Implementaion of dining philosophers problem"


def eat(index, left_fork, right_fork):
    '''Pick up left & right fork and eat'''
    left_fork.acquire()
    right_fork.acquire()
    eating_philosophers[index % 5], eating_philosophers[(index + 1) % 5] = index + 1, index + 1
    print(f"Philosopher {index + 1} started eating")
    print(eating_philosophers)
    sleep(random.randint(1, 10))

def release_fork(index, left_fork, right_fork):
    '''Drop left & right fork'''
    left_fork.release()
    right_fork.release()
    eating_philosophers[index % 5], eating_philosophers[(index + 1) % 5] = 0, 0
    print(f"Philosopher {index + 1} stopped eating")

def think(index):
    '''Start thinking'''
    print(f"Philosopher {index + 1} started thinking")
    sleep(random.randint(1, 10))
    print(f"Philosopher {index + 1} is hungry")

def main(index, left_fork, right_fork):
    while True:
        # Eat if left & right fork are not claimed
        if not left_fork.locked() and not right_fork.locked():
            eat(index, left_fork, right_fork)
            release_fork(index, left_fork, right_fork)
        think(index)            


if __name__ == "__main__":
    fork = [Lock() for x in range(5)]
    eating_philosophers = [0] * 5
    for i in range(5):
        Thread(target=main, args=(i, fork[i % 5], fork[(i + 1) % 5])).start()

File: concurrent programming/dining philosophers problem/DiningPhilosophersProblem.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.concurrent.Semaphore;

public class DiningPhilosophersProblem {

	public static int[] eatingPhilosophers = {0, 0, 0, 0, 0};
	
	public static void main(String[] args) {
		
		ArrayList<Semaphore> forks = new ArrayList<Semaphore>(
				Arrays.asList(
						new Semaphore(1),
						new Semaphore(1),
						new Semaphore(1),
						new Semaphore(1),
						new Semaphore(1)
				)
		);
		
		for(int i = 0; i < 5; i++) {
			new Thread(new Philosophers(i, forks.get(i % 5), forks.get((i + 1) % 5))).start();
		}

	}
	
	private static class Philosophers implements Runnable {
		
		private int id;
		private Semaphore leftFork;
		private Semaphore rightFork;
		
		public Philosophers(int id, Semaphore leftFork, Semaphore rightFork) {
			this.id = id;
			this.leftFork = leftFork;
			this.rightFork = rightFork;
		}

		@Override
		public void run() {
			try {
				while(true) {
					if(this.leftFork.availablePermits() == 1 && this.rightFork.availablePermits() == 1) {
						eat();
						releaseFork();
					}
					think();
				}
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		}
		
		public void eat() throws InterruptedException {
			this.leftFork.acquire();
			this.rightFork.acquire();
			System.out.println(String.format("Philosopher %d started eating", this.id + 1));
			DiningPhilosophersProblem.eatingPhilosophers[this.id % 5] = this.id + 1;
			DiningPhilosophersProblem.eatingPhilosophers[(this.id + 1) % 5] = this.id + 1;
			System.out.println(Arrays.toString(DiningPhilosophersProblem.eatingPhilosophers));
			Thread.sleep((long) (Math.random() * 10000));
		}
		
		public void releaseFork() {
			this.leftFork.release();
			this.rightFork.release();
			DiningPhilosophersProblem.eatingPhilosophers[this.id % 5] = 0;
			DiningPhilosophersProblem.eatingPhilosophers[(this.id + 1) % 5] = 0;
		}
		
		public void think() throws InterruptedException {
			System.out.println(String.format("Philosopher %d started thinking", this.id + 1));
			Thread.sleep((long) (Math.random() * 10000));
			System.out.println(String.format("Philosopher %d is hungry", this.id + 1));
		}
		
	}

}
